### PR TITLE
Add get_xattr,clean_xattr,get_first_disk_source,make_relative_path_ba…

### PR DESCRIPTION
**Pull request by chunfuwen**
…cking_files into libvirt_disk

Those methods tends to be common API for reuse by other modules

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
Below case use all added methods:

DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.relativepath.blockcommit.normal_test.network_disk.iscsi.relative_path.active_domain.shallow.top_active

**Comment by chunfuwen** [comment]
Test result can be gotten from https://github.com/autotest/tp-libvirt/pull/3384 ,also

**Comment by chunfuwen** [comment]
@smitterl 
 You are appreciated if you can help review this PR

**Comment by chunfuwen** [comment]
@kylazhang 
@dzhengfy 
please help take a look

File: virttest/utils_libvirt/libvirt_disk.py [hunk 1]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+def do_blockcommit_repeatedly(vm, device_target, options, repeated_counts):
+    """
+    Do blockcommit repeatedly
+
+    :param vm: VM instance
+    :param device_target: device target
+    :param options: blockcommit options
+    :param repeated_counts: repeated counts for executing blockcommit
+    """
+    for count in range(0, repeated_counts):
```
Comment (kamilvarga) [inline virttest/utils_libvirt/libvirt_disk.py:312]: range function starts at 0 by default, so the code can be simplified as this: "for count in range(repeated_counts):"

File: virttest/utils_libvirt/libvirt_disk.py [hunk 2]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+        virsh.blockcommit(vm.name, device_target,
+                          options, ignore_status=False, debug=True)
```
Comment (kamilvarga) [inline virttest/utils_libvirt/libvirt_disk.py:314]: What if any blockcommit fails? I see ignore_status set to False, so any issue with this command during execution should raise exception and finish execution, right? Is that expected behavior? Just curious...

File: virttest/utils_libvirt/libvirt_disk.py [hunk 3]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+
+
+def cleanup_snapshots(vm, snap_del_disks=None):
+    """
+    clean up snapshots
+
+    :param vm: VM instance
+    :param snap_del_disks: list containing snapshot files
+    """
+    snapshot_list_cmd = "virsh snapshot-list %s --tree" % vm.name
```
Comment (kamilvarga) [inline virttest/utils_libvirt/libvirt_disk.py:356]: Would be beneficial to add this command to virttest/virsh.py? Just curious...

File: virttest/utils_libvirt/libvirt_disk.py [hunk 4]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+    else:
+        root_dir = pre_set_root_dir
+        basename = origin_image
+    sub_folders = list(map(chr, range(97, 97 + back_chain_lenghth)))
+    # Make external relative path backing files.
+    backing_file_dict = collections.OrderedDict()
+    for index in range(len(sub_folders)):
+        key = sub_folders[index]
+        cmd = "mkdir -p %s" % os.path.join(root_dir, key)
+        process.run(cmd, shell=True, ignore_status=False)
```
Comment (kylazhang) [inline virttest/utils_libvirt/libvirt_disk.py:206]: I think we can use python utils to implement the mkdir

File: virttest/utils_libvirt/libvirt_disk.py [hunk 5]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+            backing_key = syslink_folder_list[index-1]
+            backing_file_dict[key] = "../%s/%s.img" % (backing_key, backing_key)
+    # Create system link backing files
+    disk_format = origin_image_format
+    for key, value in list(backing_file_dict.items()):
+        backing_file_path = os.path.join(root_dir, key)
+        cmd = ("cd %s && qemu-img create -f %s -o backing_file=%s,backing_fmt=%s %s.img"
+               % (backing_file_path, created_image_format, value, disk_format, key))
+        process.run(cmd, shell=True, ignore_status=False)
+        disk_format = created_image_format
```
Comment (kylazhang) [inline virttest/utils_libvirt/libvirt_disk.py:299]: What's the usage for this line?

File: virttest/utils_libvirt/libvirt_disk.py (hunk 3 above)
Comment (kylazhang) [inline virttest/utils_libvirt/libvirt_disk.py:356]: Could we use virsh.snapshot_list instead?

File: virttest/utils_libvirt/libvirt_disk.py [hunk 6]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+
+def cleanup_snapshots(vm, snap_del_disks=None):
+    """
+    clean up snapshots
+
+    :param vm: VM instance
+    :param snap_del_disks: list containing snapshot files
+    """
+    snapshot_list_cmd = "virsh snapshot-list %s --tree" % vm.name
+    result_output = process.run(snapshot_list_cmd,
```
Comment (kylazhang) [inline virttest/utils_libvirt/libvirt_disk.py:357]: Could we use virsh.snapshot_list instead?

File: virttest/utils_libvirt/libvirt_disk.py [hunk 7]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+
+
+def clean_images_with_xattr(dirty_images):
+    """
+    Clean the images' xattr
+
+    :param dirty_images: The image list to be cleaned
+    """
+    for image in dirty_images:
+        get_attr_cmd = "getfattr -m trusted.libvirt.security -d %s" % image
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:159]: @kamilvarga 
updated

File: virttest/utils_libvirt/libvirt_disk.py [hunk 8]
```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+    :return: absolute path of top active file and back chain files list
+    """
+    if pre_set_root_dir is None:
+        first_disk_source = get_first_disk_source(vm)
+        root_dir = os.path.dirname(first_disk_source)
+        basename = os.path.basename(first_disk_source)
+    else:
+        root_dir = pre_set_root_dir
+        basename = origin_image
+    sub_folders = list(map(chr, range(97, 97 + back_chain_lenghth)))
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:200]: this intends to generate random folder name starting with  letter 'a' with specified length by back_chain_lenghth.
And code updated to [chr(letter) for letter in range(ord('a'), ord('a') + back_chain_lenghth)]

File: virttest/utils_libvirt/libvirt_disk.py (hunk 4 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:206]: @kylazhang 
Yes, but we need ignore error when folder exists already, but python utils will throw OSError.

os.mkdir() method in Python is used to create a directory named path with the specified numeric mode. This method raise FileExistsError if the directory to be created already exists.

File: virttest/utils_libvirt/libvirt_disk.py (hunk 5 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:299]: @kylazhang 
First backchain disk format is passed in by parameter.
After that all subsequent backing disk format need set to created_image_format

File: virttest/utils_libvirt/libvirt_disk.py (hunk 2 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:314]: @kamilvarga 
Yes, that is expected behavior.
Any intermediate blockcommit failure will stop subsequent operations.

File: virttest/utils_libvirt/libvirt_disk.py (hunk 6 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:357]: @kylazhang 
Primitive virsh.snapshot-list in virsh.py hardcode output parser, which can not deal with output correctly if --tree option is used
```
def snapshot_list(name, options=None, **dargs):
    """
    Get list of snapshots of domain.

    :param name: name of domain
    :param options: options of snapshot_list
    :param dargs: standardized virsh function API keywords
    :return: list of snapshot names
    """
    # CmdResult is handled here, force ignore_status
    dargs['ignore_status'] = True
    ret = []
    cmd = "snapshot-list %s" % name
    if options is not None:
        cmd += " %s" % options

    sc_output = command(cmd, **dargs)
    if sc_output.exit_status != 0:
        raise process.CmdError(
            cmd, sc_output, "Failed to get list of snapshots")

    data = re.findall("\S* *\d*-\d*-\d* \d*:\d*:\d* [+-]\d* \w*",
                      sc_output.stdout_text)
    for rec in data:
        if not rec:
            continue
        ret.append(re.match("\S*", rec).group())

    return ret

```

File: virttest/utils_libvirt/libvirt_disk.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:200]: this intends to generate random folder name starting with letter 'a' with specified length by back_chain_lenghth.
And all codes updated to [chr(letter) for letter in range(ord('a'), ord('a') + back_chain_lenghth)] accordingly

File: virttest/utils_libvirt/libvirt_disk.py (hunk 1 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:312]: updated

File: virttest/utils_libvirt/libvirt_disk.py (hunk 3 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:356]: @kamilvarga

We already have one available primitive snapshot_list command in virsh.py.
But unfornately, it doesn't support '--tree' parser.
But '--tree' is rarely used, so how about leave it as single case this time. And we can move it to virsh.py if we see more cases coming.
By the way, it is old codes existed in tp-libvirt already, but just move it to avocado-vt

File: virttest/utils_libvirt/libvirt_disk.py (hunk 4 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:206]: @kylazhang
Yes, but sometimes we need ignore error when folder exists already, but python utils will throw OSError.

os.mkdir() method in Python is used to create a directory named path with the specified numeric mode. This method raise FileExistsError if the directory to be created already exists.

File: virttest/utils_libvirt/libvirt_disk.py (hunk 5 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:299]: @kylazhang
First backchain disk format is passed in by parameter.
After that all subsequent backing disk format need set to created_image_format

File: virttest/utils_libvirt/libvirt_disk.py (hunk 6 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:357]: @kylazhang
Primitive virsh.snapshot-list in virsh.py hardcode output parser, which can not deal with output correctly if --tree option is used

def snapshot_list(name, options=None, **dargs):
    """
    Get list of snapshots of domain.

    :param name: name of domain
    :param options: options of snapshot_list
    :param dargs: standardized virsh function API keywords
    :return: list of snapshot names
    """
    # CmdResult is handled here, force ignore_status
    dargs['ignore_status'] = True
    ret = []
    cmd = "snapshot-list %s" % name
    if options is not None:
        cmd += " %s" % options

    sc_output = command(cmd, **dargs)
    if sc_output.exit_status != 0:
        raise process.CmdError(
            cmd, sc_output, "Failed to get list of snapshots")

    data = re.findall("\S* *\d*-\d*-\d* \d*:\d*:\d* [+-]\d* \w*",
                      sc_output.stdout_text)
    for rec in data:
        if not rec:
            continue
        ret.append(re.match("\S*", rec).group())

    return ret

By the way, it is old codes existed in tp-libvirt already, but just move it to avocado-vt

**Comment by chunfuwen** [comment]
@kylazhang 
@kamilvarga 
Your comments are addressed, please double check.

**Comment by chunfuwen** [comment]
new test result after code change:
Command 'virsh list' finished with 0 after 0.48030519485473633s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.relativepath.blockcommit.normal_test.network_disk.iscsi.relative_path.active_domain.shallow.top_active

File: virttest/utils_libvirt/libvirt_disk.py [hunk 9]
```diff
@@ -119,3 +127,299 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+    :return: backing chain list
+    """
+    cmd = "qemu-img info %s --backing-chain" % disk_src_file
+    if libvirt_storage.check_qemu_image_lock_support():
+        cmd = "qemu-img info -U %s --backing-chain" % disk_src_file
+    ret = process.run(cmd, shell=True).stdout_text.strip()
+    logging.debug("The actual qemu-img output:%s\n", ret)
+    match = re.findall(r'(backing file: )(.+\n)', ret)
+    qemu_img_info_backing_chain = []
+    for i in range(0, len(match)):
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py]: @kamilvarga 
thanks
now updated

File: virttest/utils_libvirt/libvirt_disk.py (hunk 4 above)
Comment (kylazhang) [reply virttest/utils_libvirt/libvirt_disk.py:215]: @chunfuwen you can use os.makedirs(directory, exist_ok=True)

File: virttest/utils_libvirt/libvirt_disk.py [hunk 10]
```diff
@@ -119,3 +127,299 @@ def create_custom_metadata_disk(disk_path, disk_format,
...
+    """
+    root_dir = pre_set_root_dir
+    syslink_folder_list = [chr(letter) for letter in range(ord('a'), ord('a') + syslink_back_chain_lenghth)]
+    backing_file_dict = collections.OrderedDict()
+    # Make external relative path backing files.
+    for index in range(len(syslink_folder_list)):
+        key = syslink_folder_list[index]
+        folder_path = os.path.join(root_dir, key)
+        cmd = "mkdir -p %s" % folder_path
+        process.run(cmd, shell=True, ignore_status=False)
```
Comment (kylazhang) [inline virttest/utils_libvirt/libvirt_disk.py:292]: As above

File: virttest/utils_libvirt/libvirt_disk.py (hunk 5 above)
Comment (kylazhang) [reply virttest/utils_libvirt/libvirt_disk.py:308]: Thanks. And I saw some duplication between this function and make_relative_path_backing_files, could you please write new function for common parts and reduce the duplication?

File: virttest/utils_libvirt/libvirt_disk.py (hunk 6 above)
Comment (kylazhang) [reply virttest/utils_libvirt/libvirt_disk.py:367]: OK, get the reason.

File: virttest/utils_libvirt/libvirt_disk.py (hunk 4 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:215]: @kylazhang 

not anticipated as you suggest when using  os.makedirs(directory, exist_ok=True)
```
>>> import os
>>> root_dir = '/root/workdir/tp-libvirt'
>>> syslink_folder_list = ['a', 'b', 'c']
>>> for index in range(len(syslink_folder_list)):
...     key = syslink_folder_list[index]
...     folder_path = os.path.join(root_dir, key)
...     os.makedirs(folder_path, True)
... 
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "/usr/lib64/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/root/workdir/tp-libvirt/a'
```

File: virttest/utils_libvirt/libvirt_disk.py (hunk 5 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:318]: @kylazhang 
wrap those code lines into one separate method:_execute_create_backend_file().

File: virttest/utils_libvirt/libvirt_disk.py (hunk 10 above)
Comment (chunfuwen) [reply virttest/utils_libvirt/libvirt_disk.py:287]: @kylazhang

not anticipated as you suggest when using os.makedirs(directory, exist_ok=True), see previous paste code

**Comment by chunfuwen** [comment]
@smitterl 
Please help take a look if time is available to you.

**Review by kylazhang — APPROVED** [review]

**Review by kamilvarga — APPROVED** [review]
LGTM

**Pull request merged**